iped-app/src/main/java/iped/viewers/timelinegraph/cache/EventTimestampCache.java

```diff
@@ -0,0 +1,149 @@
+    }
+
+    public void run() {
+        LeafReader reader = resultsProvider.getIPEDSource().getLeafReader();
+
+        IndexTimeStampCache timeStampCache = (IndexTimeStampCache) this.timeStampCache;
+
+        DocIdSetIterator timeStampValues;
+        try {
+            String eventField = ipedChartsPanel.getTimeEventColumnName(eventType);
+            if (eventField != null) {
+                timeStampValues = reader.getSortedDocValues(eventField);
+                if (timeStampValues == null) {
+                    SortedSetDocValues values = reader.getSortedSetDocValues(eventField);
+                    ArrayList<Date> parsedDateCache = new ArrayList<>();
+                    int emptyValueOrd = -1;
+                    int doc = values.nextDoc();
+                    while (doc != DocIdSetIterator.NO_MORE_DOCS) {
+                        int ord = (int) values.nextOrd();
+                        outer: while (ord != SortedSetDocValues.NO_MORE_ORDS) {
+                            if (ord != emptyValueOrd) {
+                                for (Class<? extends TimePeriod> timePeriodClass : timeStampCache.getPeriodClassesToCache()) {
+                                    Date date = null;
+                                    if (ord < parsedDateCache.size()) {
+                                        date = parsedDateCache.get(ord);
+                                    }
+                                    if (date == null) {
+                                        String timeStr = cloneBr(values.lookupOrd(ord));
+                                        if (timeStr.isEmpty()) {
+                                            emptyValueOrd = ord;
+                                            continue outer;
+                                        }
+                                        date = DateUtil.ISO8601DateParse(timeStr);
+                                        while (ord >= parsedDateCache.size()) {
+                                            parsedDateCache.add(null);
+                                        }
+                                        parsedDateCache.set(ord, date);
+                                    }
+                                    TimePeriod t = ipedChartsPanel.getDomainAxis().getDateOnConfiguredTimePeriod(timePeriodClass, date);
+                                    if (t != null) {
+                                        ArrayList<Integer> docs2 = timeStampCache.get(timePeriodClass, t, eventType);
+                                        if (docs2 == null) {
+                                            docs2 = new ArrayList<Integer>();
+                                            synchronized (timeStampCache) {
+                                                timeStampCache.add(timePeriodClass, t, eventType, docs2);
+                                            }
+                                        }
+                                        synchronized (docs2) {
```

Comment:
Is this synchronization needed?

Reply:
I read that ArrayList list is not ThreadSafe. My intention was to avoid any changes on the array list while adding the new item.